og.install

```diff
@@ -5,10 +5,60 @@
  * Install, update, and uninstall functions for the Organic groups module.
  */
 
+use Drupal\Core\Field\BaseFieldDefinition;
+
 /**
  * Implements hook_uninstall().
  */
 function og_uninstall() {
   \Drupal::queue('og_orphaned_group_content')->deleteQueue();
   \Drupal::queue('og_orphaned_group_content_cron')->deleteQueue();
 }
+
+/**
+ * Add a base field to store the group bundle in memberships.
+ */
+function og_update_8001(&$sandbox) {
+  $storage = \Drupal::entityTypeManager()->getStorage('og_membership');
+
+  if (!isset($sandbox['total'])) {
+    $storage_definition = BaseFieldDefinition::create('string')
+      ->setLabel(t('Group bundle id'))
+      ->setDescription(t("The bundle ID of the group."));
+    \Drupal::entityDefinitionUpdateManager()->installFieldStorageDefinition('entity_bundle', 'og_membership', 'og', $storage_definition);
+
+    $sandbox['#finished'] = 0;
+    $sandbox['current'] = 0;
+    $sandbox['total'] = $storage->getQuery()->count()->execute();
+    $sandbox['batch_size'] = 500;
+  }
+
+  // Update the existing memberships to include the group bundle ID.
+  $membership_ids = $storage->getQuery()
+    ->range($sandbox['current'], $sandbox['batch_size'])
+    ->sort('id')
+    ->execute();
+
+  /** @var \Drupal\og\Entity\OgMembership $membership */
+  foreach ($storage->loadMultiple($membership_ids) as $membership) {
+    $group = $membership->getGroup();
+    if (!empty($group)) {
```

Comment:
As part of the update, we should probably already have a group (as we're not creating a new membership). Anyway, we can keep as is.

Reply:
Very good comment. I was not sure about this one. I think there are cases where there might be orphaned data left when the update hook is executed. For example if the project has configured orphans to be cleaned up on cron. I don't want the update to fail on potential missing groups, since that could be a very stressful situation for a devops engineer during a deployment.

I think it might probably be a good idea to remove the membership at this point, since a membership for a group that no longer exists is orphaned and can safely be removed. What do you think?

This comment also makes me realize there is a potential bug in the protected `OgMembership::hasGroup()` method that is being added here. If this update hook finds that the group no longer exists, then it will not update the membership, and the `entity_bundle` field will contain `NULL`. This will then blow up in `OgMembership::getRoles()`, since it will call `::hasGroup()` which doesn't currently check the presence of the bundle, and then it will try to retrieve a non-existing role.

I will address both: I will clean up the orphaned memberships as part of the update hook, and will update `::hasGroup()` to also check if the bundle is populated.